src/Compilers/Core/Portable/SourceGeneration/IncrementalGeneratorInitializationContext_ForAttributeWithMetadataName.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Xml.Serialization;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.SourceGeneration;
+using System.Threading;
+
+namespace Microsoft.CodeAnalysis;
+
+using Aliases = ArrayBuilder<(string aliasName, string symbolName)>;
+
+internal readonly struct GeneratorAttributeSyntaxContext
+{
+    /// <summary>
+    /// The syntax node the attribute is attached to.  For example, with <c>[CLSCompliant] class C { }</c> this would
+    /// the class declaration node.
+    /// </summary>
+    public SyntaxNode AttributeTarget { get; }
+
+    /// <summary>
+    /// Semantic model for the file that <see cref="AttributeTarget"/> is contained within.
+    /// </summary>
+    public SemanticModel SemanticModel { get; }
+
+    /// <summary>
+    /// <see cref="AttributeData"/>s for any matching attributes on <see cref="AttributeTarget"/>.  Always non-empty.
+    /// </summary>
+    public ImmutableArray<AttributeData> Attributes { get; }
+
+    internal GeneratorAttributeSyntaxContext(
+        SyntaxNode attribueTarget,
+        SemanticModel semanticModel,
+        ImmutableArray<AttributeData> attributes)
+    {
+        AttributeTarget = attribueTarget;
+        SemanticModel = semanticModel;
+        Attributes = attributes;
+    }
+}
+
+public partial struct IncrementalGeneratorInitializationContext
+{
+    private static readonly char[] s_nestedTypeNameSeparators = new char[] { '+' };
+    private static readonly SymbolDisplayFormat s_metadataDisplayFormat =
+        SymbolDisplayFormat.QualifiedNameArityFormat.AddCompilerInternalOptions(SymbolDisplayCompilerInternalOptions.UsePlusForNestedTypes);
+    /// <summary>
```

Comment:
@chsienki @333fred this is hte new api based on the last review.  Major changes:

1. Fixed the issue with duplicate passing of AttributeData
2. This is no longer generic on any TSyntaxNode.  
3. This follows the pattern of SyntaxNodeProvider and has a predicate/transform.
4. REmoved all overloads.

Things i didn't do:

1. I didn't add the target-symbol to the context object.  Wasn't needed yet.  Could add later.
2. Didn't use Kinds/TYpes for any sort of filtering.  Now, you just pass a predicate and do what you want.  
3.